### PR TITLE
Concurrent archives

### PR DESCRIPTION
**Pull request by thomas-zahner**
This pull request makes the archive requests concurrent using `for_each_concurrent`. Previously they were done sequentially.

### CODE REVIEW

**Comment by thomas-zahner** [comment]
@mre Should I add a new test to cli.rs for the `--suggest` flag, or maybe some other test?

File: lychee-bin/src/commands/check.rs [hunk 1]
```diff
@@ -283,6 +285,19 @@ fn show_progress(
     Ok(())
 }
 
+fn get_failed_urls(stats: &mut ResponseStats) -> Vec<(InputSource, Url)> {
+    stats
+        .fail_map
+        .iter()
+        .flat_map(|(source, set)| set.iter().map(|entry| (source, entry)).collect::<Vec<_>>())
```
[the remark this answers is not in the document]
Comment (thomas-zahner) [reply lychee-bin/src/commands/check.rs]: Thank you, I have removed the `collect` now.
But I'm not really sure how I could combine lines 292-297.
If I rewrite the lines with a for loop I would get something like the following code, but I don't know how I could further simplify it.

```rust
let mut result = Vec::new();
for (source, set) in stats.fail_map.iter() {
    for entry in set {
        let uri = &entry.uri;
        if !(uri.is_data() || uri.is_mail() || uri.is_file()) {
            result.push((source.clone(), entry.uri.as_str().try_into().unwrap()));
        }
    }
}
result
```

**Comment by mre** [comment]
> @mre Should I add a new test to cli.rs for the --suggest flag, or maybe some other test?

Since this just adds concurrency to an existing feature, I don't think we need to add any other test. We should have all code paths covered already, right?

**Comment by thomas-zahner** [comment]
Yes, this PR just extends the existing feature. But there are only tests for the wayback module.
`suggest_archived_links` is currently untested so maybe it would make sense to add a test to cli.rs?

File: lychee-bin/src/commands/check.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lebensterben) [reply lychee-bin/src/commands/check.rs]: Okay. Let me try to explain this

We have an iterator of `(InputSource, HashSet<ResponseBody>)`.

L292 transforms from this to a flattened iterator of `(InputSource, ResponseBody)`

L293-296 filters out some `ResponseBody` based on its `uri`. (`ResponseBody` is just `{ Uri, Status }`)

L297 tries to convert `Uri` into `Url`.

It can be simplified as this:

- First, what we need is only `Uri` of `ResponseBody`. So the `flat_map` on L292 can be changed accrodingly.
- Second, we can combine `filter` on L293-296 and `map` on L297 into a single `filter_map` that:
  - If the `Uri` passes the sets of conditionals, try to convert it to a `Url`.
  - Otherwise, simply yield a `None`.

So L292 becomes

```rust
        .flat_map(|(source, set)| set.iter().map(move |ResponseBody {uri, _status}| (source, uri)))
```

L293-297 becomes

```rust
        .filter_map(|(source, uri)| {
            if uri.is_data() || uri.is_mail() || uri.is_file() {
                None
            } else {
                Some((source.clone(), Url::try_from(uri.as_str()).unwrap()))
            }
        })
```

[the remark this answers is not in the document]
Comment (lebensterben) [reply lychee-bin/src/commands/check.rs]: I wrote these by hand so there might be some issues or typos.

[the remark this answers is not in the document]
Comment (mre) [reply lychee-bin/src/commands/check.rs]: @thomas-zahner, would you say that you have the bandwidth to make these changes? If not we can take over or merge it as is.

[the remark this answers is not in the document]
Comment (thomas-zahner) [reply lychee-bin/src/commands/check.rs]: I saw that you created the new version which is great. I think I will be able to continue on this at the end of this week or next week.

[the remark this answers is not in the document]
Comment (mre) [reply lychee-bin/src/commands/check.rs]: Sounds good, no rush. 😉

[the remark this answers is not in the document]
Comment (thomas-zahner) [reply lychee-bin/src/commands/check.rs]: @lebensterben Oh I see, thank you very much for the detailed explanation. I have now updated it.

@mre I have now added the CLI test and synced the branch with master, so it should be ready to be reviewed and merged.

**Comment by mre** [comment]
@thomas-zahner, I saw that you made progress on this PR and the tests pass as well. Is this ready for review or are you still working on the PR?

**Comment by thomas-zahner** [comment]
@mre Yes, I have now added the CLI test and synced the branch with master, so it is ready to be reviewed and merged.

**Review by mre — APPROVED** [review]
Looks good to me and works locally. Merging. Thanks!

**Comment by mre** [comment]
Not sure why, but the tests in `master` are flaky.
E.g. https://github.com/lycheeverse/lychee/actions/runs/4951214704/jobs/8901267952
Not sure why. Tested locally and they pass every time. Any ideas? 🤔

**Comment by thomas-zahner** [comment]
@mre Sorry for the late response, we are quite busy now with finishing up our link management tool and documenting everything.

I did some debugging on the problem. For me the tests were working quite reliably but still rarely the did fail.
I created a small bash script because the problem was so rare for me:

```bash
#!/bin/bash

substring="Suggestions"

for i in {0..1000}
do
  result=$(./target/debug/lychee fixtures/INTERNET_ARCHIVE.md --suggest)
  if [[ "$result" == *"$substring"* ]]; then
    echo "It's there."
  else
    echo "Not found!!"
    echo "$result"
    exit
  fi
done
```

The first time the script even passed for me. In the second round, roughly 1200 attempts later it failed for the first time. But no error or panic happens in the current state, only the "Suggestion" part is missing.

In check.rs I updated this:

```rust
            if let Ok(Some(suggestion)) = future.await {
                suggestions
                    .lock()
                    .unwrap()
                    .entry(input.clone())
                    .or_default()
                    .insert(Suggestion {
                        suggestion,
                        original: url.clone(),
                    });
            }
```

to

```rust
            match future.await {
                Ok(Some(suggestion)) => {
                    suggestions
                        .lock()
                        .unwrap()
                        .entry(input.clone())
                        .or_default()
                        .insert(Suggestion {
                            suggestion,
                            original: url.clone(),
                        });
                }
                Err(e) => panic!("{e}"),
                _ => {}
            }
```

Then when I managed to get it to fail a second time, lychee paniced with:

```
thread 'main' panicked at 'error sending request for url (https://archive.org/wayback/available?url=https://www.google.com/jobs.html): error trying to connect: dns error: no record found for Query { name: Name("archive.org.v.cablecom.net."), query_type: AAAA, query_class: IN }', lychee-bin/src/commands/check.rs:139:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

But this looks just like my internet provider refused to resolve the DNS entry because I was spamming it. :sweat_smile: 
My best guess is that the pipeline failed because the Wayback Machine API was unavailable or unreliable at the moment when the pipeline failed.

I could open up a new PR for handling the error similarly to the above code to get useful error messages.
If the test is really unreliable in practice we could also remove or ignore it. What do you think would be the best?
Or should we somehow mock the Wayback API in the CLI test?
Was the pipeline failing just once or did it happen multiple times?

**Comment by mre** [comment]
Wow, thanks for the thorough investigation. It seems to be a rare edge condition, so no need to fix it immediately. Although it seems the error occurs much more often on GitHub (CI) than on a local machine.

I agree that it is likely related to Wayback machine not responding or taking too long to respond. It looks like it only affects the unit tests because we unwrap on the result there. We could retry a few times maybe and only fail the test if it fails multiple times. There's no rush to change that. Let's first see if the error occurs consistently in CI.

Thanks for having a look. 😎

**Comment by thomas-zahner** [comment]
No problem, sounds good :+1:
